### DISCUSSION TITLE
XMLSec 1.3.1-rc1

### DISCUSSION THREAD

**Discussion by lsh123** [lsh123/xmlsec#659] [Announcements]
The XMLSec 1.3.1 release includes the following changes:

- Added "--with-libltdl" option for ./configure to allow custom libltdl installations and deprecated "--enable-crypto-dl" option.
- Added support for cclang compiler on non-MacOSX platforms.
- (xmlsec-nss) Bumped minimum supported NSS version to 3.52.
- Several other small fixes ([more details](https://github.com/lsh123/xmlsec/commits/master)).

Please test the [release candidate](https://www.aleksey.com/xmlsec/download/xmlsec1-1.3.1-rc1.tar.gz) ([signature](https://www.aleksey.com/xmlsec/download/xmlsec1-1.3.1-rc1.sig)) and let me know if you see any issues!

Aleksey

<hr /><em>This discussion was created from the release <a href='https://github.com/lsh123/xmlsec/releases/tag/xmlsec_1_3_1_rc1'>1.3.1-rc1</a>.</em>

**Comment by vmiklos** [comment]
Looks good on the odfsig side.

xmlsec-1.3 in libreoffice is so far not a success, but that is mainly not your fault. :-) https://bugs.documentfoundation.org/show_bug.cgi?id=155034 has more details, but in short xmlsec-1.3 uses some newer NSS symbols that seems to be not too interesting in our use-case but does cause problems when binaries are used on older distros, so at the moment we had to go back to xmlsec-1.2 for now. It's not clear to me if it makes sense to spend energy on having some --disable-whatnot switch that avoids these newer symbols or just give some time and time will fix these problems. (Just a FYI, as I feel a bit bad that you spent so much time on addressing our use-case, ans in the short-term now we don't actually use this new release due to unexpected problems.)

**Reply by lsh123** [reply]
No worries, all good. What specific symbols are missing? The "--disable-something" flags might already exists :)

**Reply by lsh123** [reply]
And what NSS do you need? 3.35?

**Reply by vmiklos** [reply]
The weird setup is that we produce binaries on RHEL7, but RHEL nicely updates NSS even in their stable versions quite frequently, so we have 3.79.0 there. Now, just by accident, these binaries also run on Ubuntu 18.04, which only has 3.35. This used to work with xmlsec-1.2, but xmlsec-1.3 has some newer NSS requirement. So what is an accident is also seen by some users as a regression if we update as-is...

Another complexity is that I think Ubuntu 18.04 is EOL by now, so possibly it's enough to get Ubuntu 20.04 working for us to update to xmlsec-1.3. Do you have any opinion on that? It's certainly an option that I check back with our bugreporter if they still care about 18.04; and if not then let's just focus on 20.04, which has 3.49.1, which is much closer to your current 3.52 baseline.

**Reply by vmiklos** [reply]
Oh and wrt which symbols, I don't have the exact error message at hand, but when do a `podman run --rm -ti ubuntu:18.04`, and lower the configure.ac versions, so configure pass with:

```diff
diff --git a/configure.ac b/configure.ac
index 721ed835..776ec6ca 100644
--- a/configure.ac
+++ b/configure.ac
@@ -951,8 +951,8 @@ dnl
 dnl ==========================================================================
 
 dnl After upgrading min ver >= 3.59, remove the check in include/nss/crypto.h (search for NSS_VMAJOR)
-NSS_MIN_VERSION="3.52"
-NSPR_MIN_VERSION="4.25.0"
+NSS_MIN_VERSION="3.35"
+NSPR_MIN_VERSION="4.18.0"
 SEAMONKEY_MIN_VERSION="1.0"
 MOZILLA_MIN_VERSION="1.4"
 NSS_CRYPTO_LIB="$XMLSEC_PACKAGE-nss"
```

then the error I get is:

```
root@a00f9cd8eb6e:~/xmlsec# make
make  all-recursive
make[1]: Entering directory '/root/xmlsec'
Making all in include
make[2]: Entering directory '/root/xmlsec/include'
Making all in xmlsec
make[3]: Entering directory '/root/xmlsec/include/xmlsec'
Making all in nss
make[4]: Entering directory '/root/xmlsec/include/xmlsec/nss'
make[4]: Nothing to be done for 'all'.
make[4]: Leaving directory '/root/xmlsec/include/xmlsec/nss'
make[4]: Entering directory '/root/xmlsec/include/xmlsec'
make[4]: Nothing to be done for 'all-am'.
make[4]: Leaving directory '/root/xmlsec/include/xmlsec'
make[3]: Leaving directory '/root/xmlsec/include/xmlsec'
make[3]: Entering directory '/root/xmlsec/include'
make[3]: Nothing to be done for 'all-am'.
make[3]: Leaving directory '/root/xmlsec/include'
make[2]: Leaving directory '/root/xmlsec/include'
Making all in src
make[2]: Entering directory '/root/xmlsec/src'
Making all in .
make[3]: Entering directory '/root/xmlsec/src'
make[3]: Nothing to be done for 'all-am'.
make[3]: Leaving directory '/root/xmlsec/src'
Making all in nss
make[3]: Entering directory '/root/xmlsec/src/nss'
  CC       libxmlsec1_nss_la-ciphers_gcm.lo
ciphers_gcm.c: In function 'xmlSecNssGcmCipherEncrypt':
ciphers_gcm.c:248:5: error: unknown type name 'CK_NSS_GCM_PARAMS'; did you mean 'CK_NSS_AEAD_PARAMS'?
     CK_NSS_GCM_PARAMS gcm_params;
     ^~~~~~~~~~~~~~~~~
     CK_NSS_AEAD_PARAMS
ciphers_gcm.c:299:15: error: request for member 'pIv' in something not a structure or union
     gcm_params.pIv      = iv;
               ^
ciphers_gcm.c:300:15: error: request for member 'ulIvLen' in something not a structure or union
     gcm_params.ulIvLen  = XMLSEC_NSS_GCM_CIPHER_IV_SIZE;
               ^
ciphers_gcm.c:301:15: error: request for member 'pAAD' in something not a structure or union
     gcm_params.pAAD     = NULL;
               ^
ciphers_gcm.c:302:15: error: request for member 'ulAADLen' in something not a structure or union
     gcm_params.ulAADLen = 0;
               ^
ciphers_gcm.c:303:15: error: request for member 'ulTagBits' in something not a structure or union
     gcm_params.ulTagBits = 8 * XMLSEC_NSS_GCM_CIPHER_TAG_SIZE;
               ^
ciphers_gcm.c: In function 'xmlSecNssGcmCipherDecrypt':
ciphers_gcm.c:336:5: error: unknown type name 'CK_NSS_GCM_PARAMS'; did you mean 'CK_NSS_AEAD_PARAMS'?
     CK_NSS_GCM_PARAMS gcm_params;
     ^~~~~~~~~~~~~~~~~
     CK_NSS_AEAD_PARAMS
ciphers_gcm.c:378:15: error: request for member 'pIv' in something not a structure or union
     gcm_params.pIv      = iv;
               ^
ciphers_gcm.c:379:15: error: request for member 'ulIvLen' in something not a structure or union
     gcm_params.ulIvLen  = XMLSEC_NSS_GCM_CIPHER_IV_SIZE;
               ^
ciphers_gcm.c:380:15: error: request for member 'pAAD' in something not a structure or union
     gcm_params.pAAD     = NULL;
               ^
ciphers_gcm.c:381:15: error: request for member 'ulAADLen' in something not a structure or union
     gcm_params.ulAADLen = 0;
               ^
ciphers_gcm.c:382:15: error: request for member 'ulTagBits' in something not a structure or union
     gcm_params.ulTagBits = 8 * XMLSEC_NSS_GCM_CIPHER_TAG_SIZE;
               ^
Makefile:650: recipe for target 'libxmlsec1_nss_la-ciphers_gcm.lo' failed
make[3]: *** [libxmlsec1_nss_la-ciphers_gcm.lo] Error 1
make[3]: Leaving directory '/root/xmlsec/src/nss'
Makefile:717: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/root/xmlsec/src'
Makefile:756: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/root/xmlsec'
Makefile:574: recipe for target 'all' failed
make: *** [all] Error 2
```

sounds like the AES GCM ciphers is not something you can disable with a configure switch?

**Reply by vmiklos** [reply]
See #663, I guess it's easier to discuss some concrete code proposal.

**Reply by lsh123** [reply]
PR #664 compiles against 3.35

**Reply by lsh123** [reply]
I merged this PR to master -- could  you please try it? as I said, the only downside is that you have to compile separately for NSS < 3.52 and NSS > 3.52 (unfortunately NSS broke the binary compatibility with this rename).

**Reply by vmiklos** [reply]
Yes, latest master builds again against that old NSS 3.35, thanks!

**Reply by lsh123** [reply]
great! anything else that you need for LO?

**Reply by vmiklos** [reply]
Not at the moment, thanks. Will try the xmlsec 1.2 -> 1.3 transition once again, after 1.3.1 is out, now that the NSS requirement is lowered.

**Reply by simo5** [reply]
> I merged this PR to master -- could you please try it? as I said, the only downside is that you have to compile separately for NSS < 3.52 and NSS > 3.52 (unfortunately NSS broke the binary compatibility with this rename).

It is mostly a build time incompatibility, binary-wise it should not cause any issues (ie if you built against an older version you should be able to run against a newer one w/o problems as NSS should have internal fallback code to accept the old structure still).

**Reply by lsh123** [reply]
I don't think NSS has the fallback. I believe this renaming in NSS broke binary compatibility for NSS.

**Reply by simo5** [reply]
That would be sad ...
